### PR TITLE
♻️ Remove typeguard dependency

### PR DESCRIPTION
**Pull request by chrisjsewell**
The `typeguard` dependency is a blocker to supporting Python 3.14 [#1566](https://github.com/useblocks/sphinx-needs/issues/1566#issuecomment-3626096968)
Currently, its only use is to validate configuration for need schemas

In this PR, we replace this functionality with validating the configuration against JSON schemas, using the existing `jsonschema-rs` dependency.
We generate these schemas via converting the `TypedDict` configuration to `pydantic` models first.
Note that we only use `pydantic` as a test dependency, to auto-generate the schemas and ensure they are up-to-date, rather than as a runtime dependency, since we don't want to introduce an additional dependency (with version constraints) that would make it more difficult for sphinx-needs to integrate into people's existing Python environments.

### CODE REVIEW

File: sphinx_needs/schema/config.py [hunk 1]
```diff
@@ -438,7 +340,7 @@ class NeedFieldsSchemaType(AllOfSchemaType):
     Intented to validate multiple fields on a need type.
     """
 
-    type: Literal["object"]
+    type: NotRequired[Literal["object"]]
```
Comment (chrisjsewell) [inline sphinx_needs/schema/config.py:343]: @ubmarco I believe this uncovered a bug here that `typeguard` did not pick up on; since `{"allOf": [...]}` is a valid `NeedFieldsSchemaType`, due to its inheritance of `AllOfSchemaType`, then `type` cannot be required

File: sphinx_needs/schema/config_utils.py
Comment (ubmarco) [inline sphinx_needs/schema/config_utils.py:82]: Please also check whether all items are dictionaries.
This `"schemas": [1, 2],` leads to 
```
      File "/home/marco/git/useblocks/sphinx-needs/sphinx_needs/schema/config_utils.py", line 99, in validate_schemas_config
        schema["idx"] = idx
        ~~~~~~^^^^^^^
    TypeError: 'int' object does not support item assignment
```

File: sphinx_needs/schema/config.py (hunk 1 above)
Comment (ubmarco) [reply sphinx_needs/schema/config.py:343]: Actually having `type` as a sibling to `allOf` is not a problem, because users can validate a given schema, but independently also the `allOf` key. It just does nothing if it has no other keys such as `properties` or `required`.

File: sphinx_needs/schema/config_utils.py
Comment (chrisjsewell) [reply sphinx_needs/schema/config_utils.py:82]: added 👍

File: sphinx_needs/schema/config.py (hunk 1 above)
Comment (chrisjsewell) [reply sphinx_needs/schema/config.py:343]: as discussed, currently there are existing tests that only contain `allOf` and thus fail validation failed.
We can change in back to required in a later PR, if we feel this is necessary, and also change the tests.
This would be an "edge case" breaking change for users though
